roomserver/internal/input/input_events.go

```diff
@@ -576,15 +582,17 @@ func (r *Inputer) processStateBefore(
 	if rejectionErr = gomatrixserverlib.Allowed(event, &stateBeforeAuth); rejectionErr != nil {
 		return
 	}
-	// Work out what the history visibility was at the time of the
+	// Work out what the history visibility/ACLs was at the time of the
 	// event.
 	for _, event := range stateBeforeEvent {
+		if event.Type() == "m.room.server_acl" && event.StateKeyEquals("") {
+			r.ACLs.OnServerACLUpdate(event)
```

Comment:
It feels very wrong to side-effect and call this in a function called `processStateBefore`. What happens if we call `processStateBefore` with an older event?